### PR TITLE
Refactor of organization auto-invite code

### PR DESCRIPTION
**Pull request by mwcraig**
Although this addresses some of the items in #170, it does resolve all of them. That should be kept open for now.

### CODE REVIEW

**Comment by mwcraig** [comment]
If there are no objections, the next step is to merge this and see what happens. It is still set to do a dry run only, i.e. no invites will be sent, but has been substantially rewritten to reduce the number of github API calls and to take the repository names on the command line instead of a json file.

File: add_contributors_to_org.py [hunk 1]
```diff
@@ -9,6 +9,236 @@
...
+                 oldest_date=None,
+                 min_invite_gap_days=365):
+        """
+        Generate automatic invitations to an organization based on merged
+        pull requests.
+
+        Parameters
+        ----------
+
+        organization: str
```
Comment (pllim) [inline add_contributors_to_org.py:26]:
```suggestion
        organization : str
```

File: add_contributors_to_org.py [hunk 2]
```diff
@@ -9,6 +9,236 @@
...
+            Name of the organization on GitHub.
+
+        token : str
+            GitHub token with administrative read/write permissions on
+            the organization. The account needs sufficient privileges to
+            access all information about organization membership
```
Comment (pllim) [inline add_contributors_to_org.py:32]:
```suggestion
            access all information about organization membership.
```

File: add_contributors_to_org.py [hunk 3]
```diff
@@ -9,6 +9,236 @@
...
+
+        verbose : bool, optional
+            If ``True``, print information about progress.
```
Comment (pllim) [inline add_contributors_to_org.py:35]:
```suggestion
            If `True`, print information about progress.
```

File: add_contributors_to_org.py [hunk 4]
```diff
@@ -9,6 +9,236 @@
...
+
+        dry_run : bool, optional
+            If ``True``, do everything *except* actually invitations. A
```
Comment (pllim) [inline add_contributors_to_org.py:38]:
```suggestion
            If `True`, do everything *except* actually invitations. A
```

File: add_contributors_to_org.py [hunk 5]
```diff
@@ -9,6 +9,236 @@
...
+        org_repo = self.github_connection.repository(self.org.login, repo)
+
+        pr_count = {}
+
+        if self.oldest_date is not None:
+            if isinstance(self.oldest_date, str):
+                oldest_date = date.fromisoformat(self.oldest_date)
+            too_old = oldest_date
+        else:
+            # Go back one year from now
```
Comment (pllim) [inline add_contributors_to_org.py:107]: Why one year?

File: add_contributors_to_org.py [hunk 6]
```diff
@@ -9,6 +9,236 @@
...
+            by the bot.
+        """
+        self.github_connection = github3.login(token=token)
+        self.org = self.github_connection.organization(organization)
+
+        # Build lists of categories we should skip invites for
+        # If the token does not have sufficient scope then the initialization will fail
+        # here.
+        self.blocked_users = [b.login for b in self.org.blocked_users()]
+        self.open_invitiation = [i.login for i in self.org.invitations()]
```
Comment (pllim) [inline add_contributors_to_org.py:63]: Typo?
```suggestion
        self.open_invitation = [i.login for i in self.org.invitations()]
```

File: add_contributors_to_org.py [hunk 7]
```diff
@@ -269,18 +311,16 @@ def main(token, args):
...
+                        help="Name of the GitHub organization to check for new "
+                             "invitees.")
+
+    parser.add_argument('repos', nargs='+',
+                        help="One or more repository in this organization "
+                             "which are to be checked for new pull requests.")
 
     parser.add_argument('--num-pr', '-n', action='store',
-                        default=2, type=int,
+                        default=1, type=int,
```
Comment (pllim) [inline add_contributors_to_org.py:323]: Is this a policy change?

**Review by pllim — COMMENTED** [review]
Overall, I have no objection to deploy this in dry-run mode to smoke out the remaining logistical oversight. Thanks!

Instead of a loop, one could also save all the authors and then do set operation to grab the "good" ones. I see you already do set operation but only at the end. But since you already tested this, it is fine to leave as-is.

Some minor comments below.

File: add_contributors_to_org.py [hunk 8]
```diff
@@ -9,6 +9,236 @@
...
+                break
+
+            author = pr.user.login
+
+            if (author in pr_count.keys()):
+                pr_count[author] += 1
+                continue
+            elif (author in self.blocked_users):
+                continue
+            elif (author in self.open_invitiation):
```
Comment (mwcraig) [inline add_contributors_to_org.py:131]:
```suggestion
            elif (author in self.open_invitation):
```

File: add_contributors_to_org.py (hunk 5 above)
Comment (mwcraig) [reply add_contributors_to_org.py:107]: Honestly, I just picked a number....

File: add_contributors_to_org.py (hunk 7 above)
Comment (mwcraig) [reply add_contributors_to_org.py:323]: There has never been a policy as far as I know -- once this is ready to launch for real I'll ask coordinated package authors for input on what they want the number to be. This bot won't add anybody to repository teams or anything, it just sends an invite to the astropy organization. Once they are in the org package maintainers can decide whether to add to a team, etc.

**Comment by pllim** [comment]
I'll let you decide when to merge.

**Comment by mwcraig** [comment]
I'm going to go ahead and merge this since `dry-run` is still on...

**Pull request merged**